src/Product/WCProductAdapter.php

```diff
@@ -176,21 +177,85 @@ protected function map_wc_general_attributes() {
 			$this->setItemGroupId( $this->parent_wc_product->get_id() );
 		}
 
+		// set product type using merchants defined product categories
+		$base_product_id      = $this->is_variation() ? $this->parent_wc_product->get_id() : $this->wc_product->get_id();
+		$product_category_ids = wc_get_product_cat_ids( $base_product_id );
+		if ( ! empty( $product_category_ids ) ) {
+			$google_product_types = self::convert_product_types( $product_category_ids );
+			do_action(
+				'woocommerce_gla_debug_message',
+				sprintf(
+					'Product category (ID: %s): %s.',
+					$base_product_id,
+					json_encode( $google_product_types )
+				),
+				__METHOD__
+			);
+			$this->setProductTypes( $google_product_types );
+		}
+
 		return $this;
 	}
 
+	/**
+	 * Covert WooCommerce product categories to product_type, which follows Google requirements:
+	 * https://support.google.com/merchants/answer/6324406?hl=en#
+	 *
+	 * @param int[] $category_ids
+	 *
+	 * @return array
+	 */
+	public static function convert_product_types( $category_ids ): array {
```

Comment:
It's probably best to make these methods (this one and `get_google_product_offer_id`) `protected` if we're not using them anywhere else. Especially given that this class isn't a utility class.

Reply:
convert_product_types and get_google_product_offer_id are also used in WCCouponAdapter, the 2 adapters share some product info conversion method to guarantee product attributes matches the coupon restrictions related to product ids or product categories. We can consider to create a utility for adapters in the followup improvements.

Reply:
Ah, okay then I would definitely vote for a utility class since that's not a functionality specific to the WCProductAdapter.